.github/workflows/build.yml

```diff
@@ -0,0 +1,96 @@
+name: Pull Sources from Manifest and build
+
+on:
+  push:
+    branches:
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+
+env:
+  CI_ONIRO_MANIFEST_URL: ${{ github.server_url }}/${{ github.repository }}
+  CI_ONIRO_MANIFEST_NAME: default.xml
+  CI_ONIRO_MANIFEST_BRANCH: ${{ github.ref_name }}
+  CI_ONIRO_RUNNER_PERSISTENT_STORAGE: /home/openharmony/persistent_storage
+
+jobs:
+  pull_sources:
+    runs-on: self-hosted
+    container:
+      image: swr.cn-south-1.myhuaweicloud.com/openharmony-docker/openharmony-docker:1.0.0
+      volumes:
+        - /home/francesco/actions-runner/persistent_storage:/home/openharmony/persistent_storage
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v2
+
+      - name: Set up git-repo
+        run: |
+          if [ ! -f /bin/repo ]; then
+            curl https://gitee.com/oschina/repo/raw/fork_flow/repo-py3 -o /bin/repo
+            chmod a+x /bin/repo
+          fi
+
+      - name: Sync mirror
+        run: |
+          # Create/maintain a mirror of all manifest repositories in runner
+          # persistent storage, and use that as reference when initializing the
+          # build repo, effectively reducing the time spent fetching git repos to
+          # almost nothing.
+          #
+          # If for some reason, the repo mirror becomes corrupted, a pipeline with
+          # $REBUILD_REPO_MIRROR set to a non-empty value will remove the entire
+          # mirror, and rebuild it from scratch.
+          #
+          # Note the use of fd 9 to hold a lock while accessing the $REPO_MIRROR
+          # path, so we can handle parallel jobs.  First we open the lock file on fd
+          # 9, then we lock it, and close/unlock it after repo sync.
+          if [ -n "$CI_ONIRO_RUNNER_PERSISTENT_STORAGE" ] ; then
+            REPO_MIRROR="$CI_ONIRO_RUNNER_PERSISTENT_STORAGE/repo-mirrors" ;
+            mkdir -p $(dirname $REPO_MIRROR) ;
+            exec 9>$REPO_MIRROR.lock ; flock 9 ;
+            if [ -n "$REBUILD_REPO_MIRROR" ] ; then
+              rm -rf "$REPO_MIRROR" ;
+            fi ;
+            if [ ! -e "$REPO_MIRROR" ] ; then
+              echo "Creating new repo mirror @ $REPO_MIRROR" ;
+              mkdir -p "$REPO_MIRROR" ;
+              pushd "$REPO_MIRROR" ;
+              repo init -u "$CI_ONIRO_MANIFEST_URL" -m "$CI_ONIRO_MANIFEST_NAME" -b "$CI_ONIRO_MANIFEST_BRANCH" --mirror ;
+            else
+              echo "Reusing repo mirror @ $REPO_MIRROR" ;
+              pushd "$REPO_MIRROR" ;
+              repo init -u "$CI_ONIRO_MANIFEST_URL" -m "$CI_ONIRO_MANIFEST_NAME" -b "$CI_ONIRO_MANIFEST_BRANCH";
```

Comment:
I somehow was expecting that we just follow up with repo sync, etc when the mirror folder does exist. Why do we need to run the full repo init again here?

Reply:
this part was taken from oniro-core ci, I believe repo init is needed here in case the manifest and/or branch is different from the one chosen when initializing the repo mirror.